nav_tui.py

```diff
@@ -0,0 +1,229 @@
+        self.voltage = 0
+        self.wrench = None
+        self.LAT = None
+        self.LON = None
+        self.ALT = None
+        # The fault values are initalized as zero (no faults)
+        self.FL_fault = 0
+        self.FR_fault = 0
+        self.BL_fault = 0
+        self.BR_fault = 0
+        self.motor_topics = ['/FL_motor', '/FR_motor', '/BL_motor', '/BR_motor']
+        # The meaning of different fault codes
+        self.fault_codes = {1: 'OVERHEAT', 2: 'OVERVOLTAGE', 4: 'UNDERVOLTAGE', 8: 'SHORT_CIRCUIT',
+                            16: 'EMERGENCY_STOP', 32: 'SEPEX_EXCITATION_FAULT', 64: 'MOSFET_FAILURE',
+                            128: 'STARTUP_CONFIG_FAULT'}
+
+        # Will try and access low/critical voltage ros parameters, if none are located default values are used
+        self.lowThreshold = rospy.get_param('battery-voltage/low', default=26)
+        self.criticalThreshold = rospy.get_param('battery-voltage/critical', default=20)
+
+    # Cb function for wrench type
+    def wrench_cb(self, string):
+        self.wrench = string.data
+
+    # Cb function for kill alarm
+    def kill_alarm_cb(self, alarm):
+        self.Alarm_name = alarm.alarm_name
+        self.raised = alarm.raised
+
+    # Cb function for the average of the battery voltages
+    def battery_monitor_cb(self, voltage):
+        self.voltage = np.round(voltage.data, 2)
+
+    # Cb function for LLA coordinates
+    def LLA_cb(self, cord):
+        self.LAT = cord.point.x
+        self.LON = cord.point.y
+        self.ALT = cord.point.z
+
+    # Cb function for each of the individual thrusters' status
+    def thruster_fault_cb(self, status, topic):
+        if topic == '/FL_motor':
+            self.FL_fault = status.fault
+        elif topic == '/FR_motor':
+            self.FR_fault = status.fault
+        elif topic == '/BL_motor':
+            self.BL_fault = status.fault
+        elif topic == '/BR_motor':
+            self.BR_fault = status.fault
+
+    # Method for converting the 8-bit int sum fault value into a list of the corresponding faults
+    def decode_fault_status(self, faults):
+        ordered_faults = [128, 64, 32, 16, 8, 4, 2, 1]
+        faults_list = []
+        for fault in ordered_faults:
+            if(faults >= fault):
+                faults_list.append(self.fault_codes[fault])
+                faults -= fault
+        return faults_list
+
+    def display(self):
+        self.panel.top()
+        self.panel.show()
+        self.window.clear()
+        # Initialization of color pairs for text display
+        color = curses.A_NORMAL
+        curses.init_pair(1, curses.COLOR_RED, curses.COLOR_BLACK)
+        curses.init_pair(2, curses.COLOR_YELLOW, curses.COLOR_BLACK)
+        curses.init_pair(3, curses.COLOR_GREEN, curses.COLOR_BLACK)
+        curses.init_pair(4, curses.COLOR_BLACK, curses.COLOR_WHITE)
+        curses.init_pair(5, curses.COLOR_YELLOW, curses.COLOR_BLACK)
+
+        while True:
```

Comment:
An infinite while loop with no wait is problematic. My screen flickers running this and it hogs a lot of cpu. Consider updating at a fixed rate (like 1 hz)

Reply:
Yeah I hadn't before noticed the cpu hit of running the tui but it's significant. I set the rate to be 2 hz and it greatly reduces the cpu load.